File: backend/app/schemas/procedures.py
"""Pydantic v2 schemas for /procurement + /procedures routers (Phase 5.1).

Locked spec from `docs/superpowers/plans/2026-06-23-phase5-zakupka.md` §5.1.
"""
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.deliveries import DeliveryOut


# ---------------------------------------------------------------------------
# Position sub-schema
# ---------------------------------------------------------------------------

class ProcedurePositionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    procedure_id: int
    source_id: Optional[int] = None
    name: str
    qty: float
    unit: Optional[str] = None
    gost_tu: Optional[str] = None
    doc_code: Optional[str] = None
    price: Optional[int] = None  # INTEGER kopecks
    delivery_id: Optional[int] = None  # NULL = «ожидает отгрузки»


# ---------------------------------------------------------------------------
# List shapes
# ---------------------------------------------------------------------------

class ProcedureListItem(BaseModel):
    id: int
    proc: Optional[str] = None
    tender_num: Optional[str] = None     # tender.num (№ заявки)
    code: str                            # parent.code (Т-67)
    title: str                           # parent.title
    mtr: Optional[str] = None            # proc.mtr ?? parent.mtr
    supplier: Optional[str] = None
    fio_zakupshchik: Optional[str] = None
    pub_start: Optional[str] = None
    pub_end: Optional[str] = None
    zagruzka: str                        # parent.zagruzka
    position_count: int
    status_zakup: Optional[str] = None
    created_at: str


class PaginatedProcedures(BaseModel):
    items: list[ProcedureListItem]
    total: int


# ---------------------------------------------------------------------------
# Detail
# ---------------------------------------------------------------------------

class ProcedureDetail(BaseModel):
    id: int
    proc: Optional[str] = None
    tender_id: int
    tender_num: Optional[str] = None
    parent_id: int
    code: str
    title: str
    mtr: Optional[str] = None
    supplier: Optional[str] = None
    fio_zakupshchik: Optional[str] = None
    pub_start: Optional[str] = None
    pub_end: Optional[str] = None
    zagruzka: str
    block: str
    status_zakup: Optional[str] = None
    # Б2 (Сопровождение):
    contract: Optional[str] = None
    fio_dogovornik: Optional[str] = None
    contract_sum: Optional[int] = None
    status_sdelki: Optional[str] = None
    status_postavki: Optional[str] = None
    srok_dd: Optional[str] = None
    plan_date: Optional[str] = None
    fakt_date: Optional[str] = None
    created_at: str
    positions: list[ProcedurePositionOut] = []
    deliveries: list[DeliveryOut] = []
    # NOTE: block_entered_at intentionally OMITTED (служебное)


# ---------------------------------------------------------------------------
# Patch
# ---------------------------------------------------------------------------

class ProcedurePatch(BaseModel):
    proc: Optional[str] = None
    tender_num: Optional[str] = None     # writes tender.num
    supplier: Optional[str] = None
    fio_zakupshchik: Optional[str] = None
    mtr: Optional[str] = None
    pub_start: Optional[str] = None
    pub_end: Optional[str] = None
    status_zakup: Optional[str] = None   # validated against dict (6 values)
    # Б2 (Сопровождение):
    contract: Optional[str] = None
    fio_dogovornik: Optional[str] = None
    contract_sum: Optional[int] = None
    status_sdelki: Optional[str] = None  # validated against dict (3 values)
    status_postavki: Optional[str] = None  # validated against 6-value enum
    srok_dd: Optional[str] = None
    plan_date: Optional[str] = None
    fakt_date: Optional[str] = None


# ---------------------------------------------------------------------------
# Phase 5.2 — split + priced positions + to-support
# ---------------------------------------------------------------------------

class SplitItem(BaseModel):
    source_position_id: int
    qty: float = Field(gt=0)


class SplitIn(BaseModel):
    positions: list[SplitItem] = Field(min_length=1)
    supplier: Optional[str] = None
    proc: Optional[str] = None      # sister's № процедуры (unique; NULL ok)
    mtr: Optional[str] = None       # sister's МТР override (else inherit source's)


class ProcedurePositionIn(BaseModel):
    name: str = Field(min_length=1)
    qty: float
    unit: Optional[str] = None
    gost_tu: Optional[str] = None
    doc_code: Optional[str] = None
    price: Optional[int] = None     # INTEGER kopecks
    source_id: Optional[int] = None # null = added by purchaser (no cap)


class ProcedurePositionPatch(BaseModel):
    name: Optional[str] = None
    qty: Optional[float] = None
    unit: Optional[str] = None
    gost_tu: Optional[str] = None
    doc_code: Optional[str] = None
    price: Optional[int] = None


__all__ = [
    "ProcedurePositionOut",
    "ProcedureListItem",
    "PaginatedProcedures",
    "ProcedureDetail",
    "ProcedurePatch",
    "SplitItem",
    "SplitIn",
    "ProcedurePositionIn",
    "ProcedurePositionPatch",
]

File: frontend/src/api/procedures.ts
import { apiFetch } from './client'
import type { DeliveryOut } from './support'

// ---- Types ----------------------------------------------------------------
// Mirrors backend schemas/procedures.py EXACTLY (see plan 5.1/5.2):
//   - proc / supplier / fio_zakupshchik / mtr / pub_start / pub_end are
//     nullable; status_zakup is nullable; price is INTEGER kopecks.
// The 10 sort keys whitelist the backend accepts (plan §123):
//   created_at | code | num | proc | supplier | status | mtr | zagruzka |
//   pub_start | pub_end

export type ProcedureSort =
  | 'created_at'
  | 'code'
  | 'num'
  | 'proc'
  | 'supplier'
  | 'status'
  | 'mtr'
  | 'zagruzka'
  | 'pub_start'
  | 'pub_end'

export type ProcedureListItem = {
  id: number
  proc: string | null
  tender_num: string | null
  code: string
  title: string
  mtr: string | null
  supplier: string | null
  fio_zakupshchik: string | null
  pub_start: string | null
  pub_end: string | null
  zagruzka: string
  position_count: number
  status_zakup: string | null
  created_at: string
}

export type PaginatedProcedures = {
  items: ProcedureListItem[]
  total: number
}

export type ProcedurePosition = {
  id: number
  procedure_id: number
  source_id: number | null
  name: string
  qty: number
  unit: string | null
  gost_tu: string | null
  doc_code: string | null
  price: number | null // INTEGER kopecks
  delivery_id: number | null // NULL = «ожидает отгрузки» (Фаза 6.2)
}

export type ProcedureDetail = {
  id: number
  proc: string | null
  tender_id: number
  tender_num: string | null
  parent_id: number
  code: string
  title: string
  mtr: string | null
  supplier: string | null
  fio_zakupshchik: string | null
  pub_start: string | null
  pub_end: string | null
  zagruzka: string
  block: string
  status_zakup: string | null
  // Б2 (Сопровождение, Фаза 6.2):
  contract: string | null
  fio_dogovornik: string | null
  contract_sum: number | null
  status_sdelki: string | null
  status_postavki: string | null
  srok_dd: string | null
  plan_date: string | null
  fakt_date: string | null
  created_at: string
  positions: ProcedurePosition[]
  deliveries: DeliveryOut[]
}

export type ProcedurePatch = {
  proc?: string | null
  tender_num?: string | null
  supplier?: string | null
  fio_zakupshchik?: string | null
  mtr?: string | null
  pub_start?: string | null
  pub_end?: string | null
  status_zakup?: string | null
  // Б2 (Сопровождение):
  contract?: string | null
  fio_dogovornik?: string | null
  contract_sum?: number | null
  status_sdelki?: string | null
  status_postavki?: string | null
  srok_dd?: string | null
  plan_date?: string | null
  fakt_date?: string | null
}

export type SplitItem = {
  source_position_id: number
  qty: number
}

export type SplitPayload = {
  positions: SplitItem[]
  supplier?: string | null
  proc?: string | null
  mtr?: string | null
}

export type ProcedurePositionInput = {
  name: string
  qty: number
  unit?: string | null
  gost_tu?: string | null
  doc_code?: string | null
  price?: number | null
  source_id?: number | null
}

export type ProcedurePositionPatch = {
  name?: string | null
  qty?: number | null
  unit?: string | null
  gost_tu?: string | null
  doc_code?: string | null
  price?: number | null
}

// ---- Helpers --------------------------------------------------------------

function buildQuery(params: Record<string, unknown>): string {
  const usp = new URLSearchParams()
  for (const [k, v] of Object.entries(params)) {
    if (v === undefined || v === null || v === '') continue
    usp.set(k, String(v))
  }
  const s = usp.toString()
  return s ? `?${s}` : ''
}

// ---- Endpoints ------------------------------------------------------------

export function listProcurements(params: {
  include_archived?: boolean
  search?: string
  sort?: ProcedureSort
  page?: number
  page_size?: number
} = {}): Promise<PaginatedProcedures> {
  return apiFetch<PaginatedProcedures>(`/procurement${buildQuery(params)}`)
}

export function getProcedure(id: number): Promise<ProcedureDetail> {
  return apiFetch<ProcedureDetail>(`/procedures/${id}`)
}

export function patchProcedure(
  id: number,
  payload: ProcedurePatch,
): Promise<ProcedureDetail> {
  return apiFetch<ProcedureDetail>(`/procedures/${id}`, {
    method: 'PATCH',
    body: payload,
  })
}

export function splitProcedure(
  id: number,
  payload: SplitPayload,
): Promise<ProcedureDetail> {
  return apiFetch<ProcedureDetail>(`/procedures/${id}/split`, {
    method: 'POST',
    body: payload,
  })
}

export function cancelProcedure(id: number): Promise<ProcedureDetail> {
  return apiFetch<ProcedureDetail>(`/procedures/${id}/cancel`, {
    method: 'POST',
  })
}

export function uncancelProcedure(id: number): Promise<ProcedureDetail> {
  return apiFetch<ProcedureDetail>(`/procedures/${id}/uncancel`, {
    method: 'POST',
  })
}

export function toSupport(id: number): Promise<ProcedureDetail> {
  return apiFetch<ProcedureDetail>(`/procedures/${id}/to-support`, {
    method: 'POST',
  })
}

export function listProcedurePositions(
  id: number,
): Promise<ProcedurePosition[]> {
  return apiFetch<ProcedurePosition[]>(`/procedures/${id}/positions`)
}

export function addProcedurePositions(
  id: number,
  positions: ProcedurePositionInput[],
): Promise<ProcedurePosition[]> {
  return apiFetch<ProcedurePosition[]>(`/procedures/${id}/positions`, {
    method: 'POST',
    body: positions,
  })
}

export function patchProcedurePosition(
  id: number,
  posId: number,
  payload: ProcedurePositionPatch,
): Promise<ProcedurePosition> {
  return apiFetch<ProcedurePosition>(
    `/procedures/${id}/positions/${posId}`,
    { method: 'PATCH', body: payload },
  )
}

export function deleteProcedurePosition(
  id: number,
  posId: number,
): Promise<{ ok: boolean }> {
  return apiFetch<{ ok: boolean }>(
    `/procedures/${id}/positions/${posId}`,
    { method: 'DELETE' },
  )
}
